src/control/server/harness_test.go

```diff
@@ -302,58 +444,178 @@ func TestServer_HarnessIOServerStart(t *testing.T) {
+
 				msClientCfg := mgmtSvcClientCfg{
 					ControlAddr:  &net.TCPAddr{},
-					AccessPoints: []string{"localhost"},
+					AccessPoints: []string{defaultAddrStr},
 				}
 				msClient := newMgmtSvcClient(context.TODO(), log, msClientCfg)
+				// create mock that implements MgmtSvcClient
+				mockMSClient := proto.NewMockMgmtSvcClient(
+					proto.MockMgmtSvcClientConfig{})
+				// store for checking calls later
+				mockMSClients[i] = mockMSClient.(*proto.MockMgmtSvcClient)
+				mockConnectFn := func(ctx context.Context, ap string,
+					tc *security.TransportConfig,
+					fn func(context.Context, mgmtpb.MgmtSvcClient) error,
+					extraDialOpts ...grpc.DialOption) error {
+
+					return fn(ctx, mockMSClient)
+				}
+				// inject fn that uses the mock client to be used on connect
+				msClient.connectFn = mockConnectFn
+
 				srv := NewIOServerInstance(log, bdevProvider, scmProvider, msClient, runner)
+				var isAP bool
+				if tc.isAP && i == 0 { // first instance will be AP & bootstrap MS
+					isAP = true
+				}
+				var uuid string
+				if UUID, exists := tc.instanceUuids[i]; exists {
+					uuid = UUID
+				}
+				var rank *system.Rank
+				if tc.rankInSuperblock {
+					rank = system.NewRankPtr(uint32(i + 1))
+				} else if isAP { // bootstrap will assume rank 0
+					rank = new(system.Rank)
+				}
+				srv.setSuperblock(&Superblock{
+					MS: isAP, UUID: uuid, Rank: rank, CreateMS: isAP, BootstrapMS: isAP,
+				})
+
 				if err := harness.AddInstance(srv); err != nil {
 					t.Fatal(err)
 				}
 			}
 
-			if err := harness.CreateSuperblocks(false); err != nil {
-				t.Fatal(err)
-			}
-
+			// set mock dRPC client to record call details
 			for _, srv := range harness.Instances() {
-				// simulate ready notification
 				srv.setDrpcClient(newMockDrpcClient(&mockDrpcClientConfig{
 					SendMsgResponse: &drpc.Response{},
 				}))
 			}
 
+			ctx, cancel := context.WithTimeout(context.Background(), testShortTimeout*10)
+			defer cancel()
+
+			// start harness async and signal completion
+			var gotErr error
+			membership := system.NewMembership(log)
 			done := make(chan struct{})
-			ctx, shutdown := context.WithCancel(context.Background())
-			go func(t *testing.T, expStartErr error, th *IOServerHarness) {
-				common.CmpErr(t, expStartErr, th.Start(ctx, nil, nil))
+			go func() {
+				gotErr = harness.Start(ctx, membership, nil)
 				close(done)
-			}(t, tc.expStartErr, harness)
+			}()
+
+			waitDrpcReady := make(chan struct{})
+			go func() {
+				for {
+					ready := true
+					for _, srv := range harness.Instances() {
+						if srv.waitDrpc.IsFalse() {
+							ready = false
+						}
+					}
+					if ready {
+						close(waitDrpcReady)
+						break
+					}
+					<-time.After(testShortTimeout)
+				}
+			}()
+
+			select {
+			case <-waitDrpcReady:
+			case <-ctx.Done():
+				if tc.expStartErr != context.DeadlineExceeded {
+					CmpErr(t, tc.expStartErr, gotErr)
```

Comment:
The race condition you're seeing has to do with this line. I think you need to wait on the "done" channel in this case, before looking at gotErr.